components/script/stylesheet_loader.rs

```diff
@@ -119,6 +124,118 @@ impl StylesheetContext {
+    //
+    // TODO(mrobinson): Shouldn't we also exit early if this is an import that was originally
+    // imported from a `<link>` element that has advanced a generation as well?
+    fn is_for_a_previous_generation(&self, element: &HTMLElement) -> bool {
+        if !matches!(&self.source, StylesheetContextSource::LinkElement) {
+            return false;
+        }
+        let link = element.downcast::<HTMLLinkElement>().unwrap();
+        !self
+            .request_generation_id
+            .is_none_or(|generation| generation == link.get_request_generation_id())
+    }
+
+    fn decrement_load_and_render_blockers(&self, owner: &dyn StylesheetOwner, document: &Document) {
+        if owner.parser_inserted() {
+            document.decrement_script_blocking_stylesheet_count();
+        }
+
+        // From <https://html.spec.whatwg.org/multipage/#link-type-stylesheet>:
+        // > A link element of this type is implicitly potentially render-blocking if the element
+        // > was created by its node document's parser.
+        if matches!(self.source, StylesheetContextSource::LinkElement) && owner.parser_inserted() {
+            document.decrement_render_blocking_element_count();
+        }
+    }
+
+    fn do_post_parse_tasks(self, successful: bool, stylesheet: Option<Arc<Stylesheet>>) {
+        let element = self.element.root();
+        let document = self.document.root();
+        let owner = element
+            .upcast::<Element>()
+            .as_stylesheet_owner()
+            .expect("Stylesheet not loaded by <style> or <link> element!");
+
+        if self.is_for_a_previous_generation(&element) {
+            self.decrement_load_and_render_blockers(owner, &document);
+        }
+
+        if let Some(stylesheet) = stylesheet {
+            match &self.source {
+                StylesheetContextSource::LinkElement => {
+                    let link = element.downcast::<HTMLLinkElement>().unwrap();
+                    if link.is_effectively_disabled() {
+                        stylesheet.set_disabled(true);
+                    }
+                    link.set_stylesheet(stylesheet);
+                },
+                StylesheetContextSource::Import { import_rule } => {
+                    // Layout knows about this stylesheet, because Stylo added it to the Stylist,
+                    // but Layout doesn't know about any new web fonts that it contains.
+                    document.load_web_fonts_from_stylesheet(&stylesheet);
+
+                    let mut guard = document.style_shared_lock().write();
+                    import_rule.write_with(&mut guard).stylesheet = ImportSheet::Sheet(stylesheet);
+                },
+            }
+        }
+
+        if let Some(ref shadow_root) = self.shadow_root {
+            shadow_root.root().invalidate_stylesheets();
+        } else {
+            document.invalidate_stylesheets();
+        }
+
+        self.decrement_load_and_render_blockers(owner, &document);
```

Comment:
If we don't early return above, then why do we decrement twice here?

Reply:
👍